src/Shared/FileMatcher.cs

```diff
@@ -705,128 +826,183 @@ private static void GetFilesRecursive
                         //  The exclude search will match ALL filenames OR
                         (searchToExclude.SearchData.Filespec == "*" || searchToExclude.SearchData.Filespec == "*.*" ||
                             //  The exclude search filename pattern matches the include search's pattern
-                            (searchToExclude.SearchData.Filespec == recursionState.SearchData.Filespec &&
-                            searchToExclude.SearchData.ExtensionLengthToEnforce == recursionState.SearchData.ExtensionLengthToEnforce)))
+                            searchToExclude.SearchData.Filespec == recursionState.SearchData.Filespec))
                     {
                         //  We won't get any results from this search that we would end up keeping
                         return;
                     }
                 }
             }
 
-            RecursiveStepResult nextStep = GetFilesRecursiveStep(
-                recursionState,
-                projectDirectory,
-                stripProjectDirectory,
-                getFileSystemEntries);
+            RecursiveStepResult nextStep = GetFilesRecursiveStep(recursionState);
 
-            RecursiveStepResult[] excludeNextSteps = null;
-            if (searchesToExclude != null)
+            IList<string> files = null;
+            foreach (string file in GetFilesForStep(nextStep, recursionState, projectDirectory,
+                stripProjectDirectory, getFileSystemEntries))
             {
-                excludeNextSteps = new RecursiveStepResult[searchesToExclude.Count];
-                for (int i = 0; i < searchesToExclude.Count; i++)
+                if (excludeNextSteps == null)
                 {
-                    excludeNextSteps[i] = GetFilesRecursiveStep(
-                        searchesToExclude[i],
-                        projectDirectory,
-                        stripProjectDirectory,
-                        getFileSystemEntries);
+                    files = files ?? new List<string>();
```

Comment:
you can adjust the logic here slightly to prevent the duplicated code (line 845-847) and (line 861-862)

Reply:
[Done](https://github.com/Microsoft/msbuild/pull/2572/commits/b5e4a47c725f3dbbec16b3fda6af9fd69175ea58)